File: ios/ExpoArgon2Module.swift
import ExpoModulesCore
import Argon2Swift

enum SaltEncoding: String, Enumerable {
    case hex
    case utf8
    case base64
}

enum Argon2Mode: String, Enumerable {
    case argon2i
    case argon2d
    case argon2id
    
    func getArgon2Mode() -> Argon2Type {
        switch self {
        case .argon2d:
          return .d
        case .argon2i:
          return .i
        case .argon2id:
          return .id
        }
      }
}

struct Argon2Config: Record {
    @Field
    var iterations: Int = 2
    
    @Field
    var memory: Int = 32 * 1024
    
    @Field
    var parallelism: Int = 1
    
    @Field
    var hashLength: Int = 32
    
    @Field
    var saltEncoding: SaltEncoding = .utf8
    
    @Field
    var mode: Argon2Mode = .argon2id
}

public class ExpoArgon2Module: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoArgon2")

      AsyncFunction("hashAsync") { (password: Either<String, Data>, salt: String, config: Argon2Config) in
          let saltData: Data? = switch config.saltEncoding {
          case .hex: Data(hexEncoded: salt)
          case .base64: Data(base64Encoded: salt)
          case .utf8: salt.data(using: .utf8)
          }
          
          guard let validSaltData = saltData else {
              throw InvalidSaltFormatException()
          }
          let saltObject = Salt(bytes: validSaltData)
          
          do {
            let result = if let password: String = password.get() {
              try Argon2Swift.hashPasswordString(
                password: password,
                salt: saltObject,
                iterations: config.iterations,
                memory: config.memory,
                parallelism: config.parallelism,
                length: config.hashLength,
                type: config.mode.getArgon2Mode(),
              )
            } else if let password: Data = password.get() {
              try Argon2Swift.hashPasswordBytes(
                password: password,
                salt: saltObject,
                iterations: config.iterations,
                memory: config.memory,
                parallelism: config.parallelism,
                length: config.hashLength,
                type: config.mode.getArgon2Mode(),
              )
            } else {
              throw InvalidDataFormatException()
            }
              
              let encodedHash = result.encodedString()
              let rawHash = result.hexString()
              
              let resultDictionary = [
                "rawHash": rawHash,
                "encodedHash": encodedHash,
              ]
              return resultDictionary
              
          } catch {
              throw Argon2FailedException(error.localizedDescription)
          }
      }
  }
}

extension Data {
    init?(hexEncoded: String) {
        let hex = hexEncoded.replacingOccurrences(of: "0x", with: "")
        guard hex.count.isMultiple(of: 2) else {
            return nil
        }
        
        let chars = hex.map { $0 }
        let bytes = stride(from: 0, to: chars.count, by: 2)
            .map { String(chars[$0]) + String(chars[$0 + 1]) }
            .compactMap { UInt8($0, radix: 16) }
        
        guard hex.count / bytes.count == 2 else { return nil }
        self.init(bytes)
    }
}

final class InvalidSaltFormatException : Exception, @unchecked Sendable {
    override var reason: String {
        "Invalid salt format"
    }
}

final class InvalidDataFormatException : Exception, @unchecked Sendable {
  override var reason: String {
    "Invalid password data format"
  }
}

final class Argon2FailedException : GenericException<String>, @unchecked Sendable {
    override var reason: String {
        "Failed to generate Argon2 hash: \(param)"
    }
}

File: android/src/main/java/expo/community/modules/argon2/ExpoArgon2Module.kt
package expo.community.modules.argon2

import android.util.Base64
import com.lambdapioneer.argon2kt.Argon2Kt
import com.lambdapioneer.argon2kt.Argon2KtResult
import expo.modules.core.errors.InvalidArgumentException
import expo.modules.kotlin.apifeatures.EitherType
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import expo.modules.kotlin.types.Either
import expo.modules.kotlin.types.Enumerable


enum class SaltEncoding(val encoding: String) : Enumerable {
  HEX("hex"),
  UTF8("utf8"),
  BASE64("base64")
}

enum class Argon2Mode(val mode: String) : Enumerable {
  ARGON2_I("argon2i"),
  ARGON2_D("argon2d"),
  ARGON2_ID("argon2id");

  fun getArgon2Mode(): com.lambdapioneer.argon2kt.Argon2Mode =
    when (this) {
      ARGON2_I -> com.lambdapioneer.argon2kt.Argon2Mode.ARGON2_I
      ARGON2_D -> com.lambdapioneer.argon2kt.Argon2Mode.ARGON2_D
      ARGON2_ID -> com.lambdapioneer.argon2kt.Argon2Mode.ARGON2_ID
    }
}

class Argon2Config: Record {
  @Field
  val iterations: Int = 2

  @Field
  val memory: Int = 32 * 1024

  @Field
  val parallelism: Int = 1

  @Field
  val hashLength: Int = 32

  @Field
  val saltEncoding: SaltEncoding = SaltEncoding.UTF8

  @Field
  val mode: Argon2Mode = Argon2Mode.ARGON2_ID
}

class ExpoArgon2Module : Module() {
  private val argon2Kt: Argon2Kt by lazy { Argon2Kt() }

  @OptIn(ExperimentalStdlibApi::class, EitherType::class)
  override fun definition() = ModuleDefinition {
    Name("ExpoArgon2")

    AsyncFunction("hashAsync") { password: Either<String, ByteArray>, salt: String, config: Argon2Config ->
      try {
        val passwordBytes: ByteArray = when {
          password.`is`(String::class) -> password.get(String::class).toByteArray(Charsets.UTF_8)
          password.`is`(ByteArray::class) -> password.get(ByteArray::class)
          else -> throw Exceptions.IllegalArgument("Invalid password data format")
        }

        val saltBytes: ByteArray = when (config.saltEncoding) {
          SaltEncoding.UTF8 -> salt.toByteArray(Charsets.UTF_8)
          SaltEncoding.BASE64 -> Base64.decode(salt, Base64.NO_WRAP)
          SaltEncoding.HEX -> salt
            .lowercase()
            .substringAfter("0x")
            .hexToByteArray(HexFormat.Default)
        }

        val hashResult: Argon2KtResult = argon2Kt.hash(
          config.mode.getArgon2Mode(),
          passwordBytes,
          saltBytes,
          config.iterations,
          config.memory,
          config.parallelism,
          config.hashLength
        )
        val rawHash = hashResult.rawHashAsHexadecimal(false)
        val encodedHash = hashResult.encodedOutputAsString()

        return@AsyncFunction mapOf(
          "rawHash" to rawHash,
          "encodedHash" to encodedHash
        )
      } catch (exception: Exception) {
        throw CodedException("Failed to generate argon2 hash: ${exception.message}", exception)
      }
    }
  }
}
